### PR TITLE
campaigns: use user tokens when reconciling changesets

### LINKED ISSUE #14989: campaigns: use tokens from user_external_accounts when reconciling campaigns
For now, we'll want to retain the fallback to the code host token for admins to retain functionality until the migration to using user tokens is complete, but regular users should either use `user_external_accounts` or fail.

### PR DESCRIPTION
**Pull request by LawnGnome**
This closes #14989, and replaces #15301.

I've implemented the approach suggested by @mrnugget in #15301: instead of making `Sourcer` an interface, we'll instead add an optional `UserSource` interface that `Source` instances can implement by adding a `WithAuthenticator` method that accepts (unsurprisingly) an `auth.Authenticator`. At present, GitHub and GitLab support this.

I've tested this against GitHub and GitLab with `auth.OAuthBearerToken` user credentials, and so far, so good. The admin fallback also still appears to work just fine.

The non-admin case also works just fine:

image

This PR does not plumb the token through to gitserver, but that's coming soon...

### CODE REVIEW

**Review by DboyRedd — APPROVED** [review]
https://github.com/sourcegraph/sourcegraph/pull/15552#issue-517063977

File: cmd/repo-updater/repos/github.go [hunk 1]
```diff
@@ -44,6 +44,11 @@ type GithubSource struct {
 	originalHostname string
 }
 
+var _ Source = &GithubSource{}
+var _ UserSource = &GithubSource{}
+var _ DraftChangesetSource = &GithubSource{}
+var _ ChangesetSource = &GithubSource{}
```
Comment (eseliger) [inline cmd/repo-updater/repos/github.go:50]: This is a duplicate of l.182 I think

File: cmd/repo-updater/repos/gitlab.go [hunk 2]
```diff
@@ -28,9 +29,15 @@ type GitLabSource struct {
...
 	baseURL             *url.URL // URL with path /api/v4 (no trailing slash)
 	nameTransformations reposource.NameTransformations
+	provider            *gitlab.ClientProvider
 	client              *gitlab.Client
 }
 
+var _ Source = &GitLabSource{}
+var _ UserSource = &GitLabSource{}
+var _ DraftChangesetSource = &GitLabSource{}
+var _ ChangesetSource = &GitLabSource{}
```
Comment (eseliger) [inline cmd/repo-updater/repos/gitlab.go:39]: I think those we had already somewhere

File: cmd/repo-updater/repos/sources.go [hunk 3]
```diff
@@ -84,6 +85,16 @@ type Source interface {
 	ExternalServices() ExternalServices
 }
 
+// A UserSource is a source that can use a custom authenticator (such as one
+// contained in a user credential) to interact with the code host, rather than
+// global credentials.
+type UserSource interface {
```
Comment (eseliger) [inline cmd/repo-updater/repos/sources.go:91]: To me this name sounds a little like "Source to get Users" (which we also probably have somewhere for authz)

File: enterprise/internal/campaigns/reconciler.go [hunk 4]
```diff
@@ -796,6 +860,19 @@ func loadChangesetSpecs(ctx context.Context, tx *Store, ch *campaigns.Changeset)
 	return
 }
 
+func loadUser(ctx context.Context, id int32) (*types.User, error) {
+	return db.Users.GetByID(ctx, id)
+}
```
Comment (eseliger) [inline enterprise/internal/campaigns/reconciler.go:865]: Given it's simplicity and ease of understanding the function by just reading the body, I would inline it but no blocker

File: enterprise/internal/campaigns/reconciler.go [hunk 5]
```diff
@@ -230,6 +238,18 @@ func (e *executor) buildChangesetSource(repo *repos.Repo, extSvc *repos.External
 		return nil, errors.New("invalid number of sources for external service")
 	}
 	src := sources[0]
+
+	if auth != nil {
```
Comment (eseliger) [inline enterprise/internal/campaigns/reconciler.go:242]:
```suggestion
        // Auth == nil means we don't have a credential, but the user is site-admin, so we want to fall back to the global token and continue.
	if auth != nil {
```

File: enterprise/internal/campaigns/reconciler_test.go [hunk 6]
```diff
@@ -1289,6 +1291,141 @@ func TestNamespaceURL(t *testing.T) {
...
+	}
+
+	rs, _ := createTestRepos(t, ctx, dbconn.Global, 1)
+	repo := rs[0]
+
+	campaignSpec := createCampaignSpec(t, ctx, store, "reconciler-test-campaign", admin.ID)
+	adminCampaign := createCampaign(t, ctx, store, "reconciler-test-campaign", admin.ID, campaignSpec.ID)
+	userCampaign := createCampaign(t, ctx, store, "reconciler-test-campaign", user.ID, campaignSpec.ID)
+
+	t.Run("imported changeset", func(t *testing.T) {
```
Comment (eseliger) [inline enterprise/internal/campaigns/reconciler_test.go:1321]:
```suggestion
	t.Run("imported changeset uses global token", func(t *testing.T) {
```

**Review by eseliger — APPROVED** [review]
LGTM! nice, simple and clean 🌟 

Q: What would it take to do that for bitbucket server? I imagine it would be worthwhile adding when it's as simple as for github and gitlab. Otherwise, documentation will be hard and it'll be basically "github and gitlab for normal users, BBS only site-admin, and if you target bbs ALSO, it will always fail"

File: enterprise/internal/campaigns/reconciler.go [hunk 7]
```diff
@@ -238,6 +258,50 @@ func (e *executor) buildChangesetSource(repo *repos.Repo, extSvc *repos.External
...
+				// not, then we need to error out.
+				user, err := loadUser(ctx, campaign.LastApplierID)
+				if err != nil {
+					return nil, errors.Wrap(err, "failed to load user applying the campaign")
+				}
+
+				if user.SiteAdmin {
+					return nil, nil
+				} else {
+					return nil, errors.Errorf("user does not have a valid credential for repo %q", e.repo.Name)
```
Comment (eseliger) [inline enterprise/internal/campaigns/reconciler.go:291]: this error should be caught and considered terminal

File: cmd/repo-updater/repos/github.go (hunk 1 above)
Comment (mrnugget) [reply cmd/repo-updater/repos/github.go:50]: Fixed in 49f72b2626

File: cmd/repo-updater/repos/gitlab.go (hunk 2 above)
Comment (mrnugget) [reply cmd/repo-updater/repos/gitlab.go:39]: Fixed in efbb082e42

File: enterprise/internal/campaigns/reconciler.go (hunk 5 above)
Comment (mrnugget) [reply enterprise/internal/campaigns/reconciler.go:242]: Fixed in 19d8dcc1d6

File: enterprise/internal/campaigns/reconciler.go (hunk 7 above)
Comment (mrnugget) [reply enterprise/internal/campaigns/reconciler.go]: I'm not yet 100% convinced that it should be a terminal error (requires the user to reapply after setting credentials vs. set credentials and see the next retry pick them up), but I implemented it like that in 6a4c687fe2

**Comment by mrnugget** [comment]
> Q: What would it take to do that for bitbucket server? I imagine it would be worthwhile adding when it's as simple as for github and gitlab. Otherwise, documentation will be hard and it'll be basically "github and gitlab for normal users, BBS only site-admin, and if you target bbs ALSO, it will always fail"

From what I can tell:

1. Allow users to put in username/password in UI
2. Send that "type" over the API, save as `auth.BasicAuth` with username/password
3. Add `WithAuthenticator` to BitbucketSource (and method on Bitbucket Server Client)
4. Extend switch statement in reconciler/etc.

**Review by mrnugget — APPROVED** [review]
Looks good! Removed a bit of unneeded code in the tests and incorporated Erik's feedback (as far as I could, not sure about the naming suggestion myself). Going to merge this now.

**Pull request merged**